### PR TITLE
test: Add testing for FileCopyMethod

### PR DESCRIPTION
**Pull request by eth-p**
~~**This commit is stacked on top of #164. That must be reviewed and merged first.**~~

## Reason

To prevent regressions, we want good test coverage no matter the `FileCopyMethod` used. We also want to make sure other `FileCopyMethod`s behave like `CopyBytes` when successful for consistency.

Instead of writing many tests to check the behaviour is the same, we re-use existing tests to check that other `FileCopyMethod`s work like `CopyBytes`. The CI runs then runs these tests so the developer does not need different operating systems and the right filesystems installed.

## Details

This series of commits changes the CI and testing so that:

 - The default `FileCopyMethod` can be changed **just in tests** using an environment variable.
 - The CI has more OS variants for using specific filesystems.
    - e.g. `btrfs` for testing copy-on-write (aka reflink)
 - The CI can choose which `FileCopyMethod` to use.

## Example

[This](https://github.com/eth-p/forked-copy/actions/runs/10543146003/job/29210759076) was a run to ensure the copy-on-write (reflink) implementation works.

Previous runs helped me catch a couple edge cases:

 - `opt.FS != nil`
 - `opt.WrapReader != nil`

That would not have been as easy without these changes.

### CODE REVIEW

**Comment by otiai10** [comment]
@eth-p Is this ready for review?

**Comment by eth-p** [comment]
Yep!

**Comment by otiai10** [comment]
Could you rebase your branch first so that only the meaning full diff will show up?

**Comment by eth-p** [comment]
> Could you rebase your branch first so that only the meaning full diff will show up?

Done

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -32,7 +41,46 @@ jobs:
     - name: Build
       run: go build -v .
 
+    - name: Set up testing environment
```
Comment (otiai10) [inline .github/workflows/go.yml:44]: Please separate yaml file and let's keep this basic test simple as it is.

Testing is not only to test our source code, but it is also a working example and source of confidence users can use this package on there environment.

When we describe so much steps to setup environment, this could be a message to users, such as, "You need to get this much preparation done to just get started"

File: all_test.go [hunk 2]
```diff
@@ -351,6 +362,9 @@ func TestOptions_PreserveOwner(t *testing.T) {
 }
 
 func TestOptions_CopyRateLimit(t *testing.T) {
+	if defaultCopyMethodName != "CopyBytes" {
+		t.Skipf("Can only test WrapReader with CopyBytes method, not %s", defaultCopyMethodName)
+	}
```
Comment (otiai10) [inline all_test.go:367]: This doesn't seem the exception of test cases, but it is an expected behavior. In other words, this package should return error if users try to use `CopyRateLimit` when invalid copy-method is set.

File: all_test.go [hunk 3]
```diff
@@ -417,6 +431,10 @@ func TestOptions_OnFileError(t *testing.T) {
 }
 
 func TestOptions_FS(t *testing.T) {
+	if defaultCopyMethodName != "CopyBytes" {
+		t.Skipf("Can only test FS reading with CopyBytes method, not %s", defaultCopyMethodName)
+	}
```
Comment (otiai10) [inline all_test.go:436]: Same

File: all_test.go [hunk 4]
```diff
@@ -17,8 +17,19 @@ import (
 //go:embed test/data/case18/assets
 var assets embed.FS
 
+func setupFileCopyMethod(m *testing.M) {
+	// Allow running all the tests with a different FileCopyMethod.
+	// We want to be able to have full coverage no matter the method.
```
Comment (otiai10) [inline all_test.go:22]: Regarding you're using `t.Skip`, it is not the "coverage" you mean. Cases which should return error should return error. Users can avoid the case, but we should not skip the test. Please do not skip.

**Review by otiai10 — CHANGES_REQUESTED** [review]

File: options.go [hunk 5]
```diff
@@ -134,6 +134,11 @@ type FileCopyMethod struct {
 	fcopy func(src, dest string, info os.FileInfo, opt Options) (err error, skipFile bool)
 }
 
+// The default FileCopyMethod.
+// This only is changed during tests.
+var defaultCopyMethod = CopyBytes
+var defaultCopyMethodName = "CopyBytes"
```
[the remark this answers is not in the document]
Comment (eth-p) [reply options.go:140]: I want to use the existing tests to check that all `FileCopyMethod`s behave the same way where possible. There isn't any other reasonable way to do this, sadly. Other possible ways were:

1. Add a new `copy.SetDefaultFileCopyMethod(method FileCopyMethod)` function.
   - Bad — Mutable global variables are bad.
 
2. Copy all the tests again, but setting a different `FileCopyMethod` for each
   - Bad — DRY
   - Bad — Multiple copies = easier to forget to change all of them

3. Change all the tests to set the `FileCopyMethod` in options using a global variable defined in `all_test.go`
   - Good — No special code for testing.
   - Bad — New tests have to be set up to include `FileCopyMethod`
      - Easy to forget to do
      - If happens, will break goal of running tests with all `FileCopyMethod`s.

Option 3 is the closest, but I didn't feel that the risk of missing test coverage was a great trade off to reduce 3 extra lines of test-only code.

File: .github/workflows/go.yml (hunk 1 above)
Comment (eth-p) [reply .github/workflows/go.yml:44]: > Please separate yaml file and let's keep this basic test simple as it is.

As a new workflow, or running a composite action in the `go.yml` workflow?

Example of how composite action would be used:

```diff
      - name: Set up Go
        uses: actions/setup-go@v4
        with:
          go-version: ${{ matrix.go }}
        id: go
      - name: Check out code into the Go module directory
        uses: actions/checkout@v3
      - name: Get dependencies
        run: go get -v -t -d ./...
      - name: Build
        run: go build -v .
+     - name: Set up testing environment
+        uses: ./.github/actions/setup-filesystem
+        id: testenv
      - name: Test
+       working-directory: ${{ testenv.outputs.working_dir }}
+       env:
+         TEST_FILESYSTEM: ${{ matrix.os.filesystem }}
+         TEST_FILECOPYMETHOD: ${{ matrix.os.copymethod }}
        run: go test -v --tags=go${{ matrix.go }}
```

Comment (otiai10) [reply .github/workflows/go.yml:44]: New workflow is better

File: all_test.go (hunk 2 above)
Comment (eth-p) [reply all_test.go:367]: > This package should return error if users try to use CopyRateLimit when invalid copy-method is set.

Yes.

I should clarify: With this pull request, I'm re-uing tests made for `CopyBytes` to check that from the user perspective, other `FileCopyMethod`s behave like `CopyBytes` where possible and do not have incorrect behaviour.

This and the `TestOptions_FS` tests are skipped because the package does return an error, which is expected behaviour. They need to be handled specially or else `go test` would always fail.

Instead of skipping, do you want me to check that they correctly return the error?

File: all_test.go (hunk 4 above)
Comment (eth-p) [reply all_test.go:22]: I'll change the `t.Skip` to check for an error instead, if that is preferred?

File: all_test.go (hunk 2 above)
Comment (otiai10) [reply all_test.go:367]: First of all, please do not say sorry, nor apologize. You're doing great contribution and I really appreciate your effort and passion. Being honest and frank is very important element to keep OSS clean.

> correctly return the error

Well, yes, if it should do so, we should do so.

File: all_test.go (hunk 4 above)
Comment (otiai10) [reply all_test.go:22]: If it should return an error, let's do so.

**Comment by otiai10** [comment]
Overall, let's use `CopyBytes` **always in all platform** by default.

I prefer simpleness and customizability, over implicit kindness.

**Comment by eth-p** [comment]
> Overall, let's use CopyBytes always in all platform by default.

To clarify, this is the still default after the commits I made. I don't intend to change it, and I don't think it would be a good idea to let users of the `copy` package change the default.

I added the `var defaultCopyMethod` and left a comment that it's only for tests because I thought it was the simplest and safest option to meet the goals of this pull request. Please let me know if there is a better option I missed, but I could not find a way to re-use the `CopyBytes` tests without giving up something. It was either:

 - Reduce maintainability — by having duplicate tests with different `FileCopyMethods`
 - Reduce simplicity of test code — by always adding the `FileCopyMethod` option to every call of `Copy`
 - Add test noise to non-test code — what I did here with the `defaultCopyMethod`

The other option would be to not re-use the `CopyBytes` tests, but that has different problems:
 - Edge cases tested in `CopyBytes` will not have tests for other `FileCopyMethods`
 - Tests can not catch when an other `FileCopyMethod` has an unexpected difference from `CopyBytes`

File: all_test.go [hunk 6]
```diff
@@ -17,12 +17,15 @@ import (
...
 
+var supportsWrapReaderOption = true
+
 func setupFileCopyMethod(m *testing.M) {
 	// Allow running all the tests with a different FileCopyMethod.
 	// We want to be able to have full coverage no matter the method.
 	switch os.Getenv("TEST_FILECOPYMETHOD") {
 	case "CopyBytes":
 		defaultCopyMethod = CopyBytes
+		supportsWrapReaderOption = true
```
Comment (eth-p) [inline all_test.go:29]: I could remove this line instead of setting it to `true` again, but I like that it's explicit about what `CopyBytes` supports.

File: .github/workflows/go.yml (hunk 1 above)
Comment (eth-p) [reply .github/workflows/go.yml]: Added.

**Comment by eth-p** [comment]
I made the requested changes.

I also made a minor improvement by removing the `defaultCopyMethodName` variable and replacing it with variables in `all_test.go` which describe the capabilities of the current "default" `FileCopyMethod`.

File: all_test.go [hunk 7]
```diff
@@ -372,8 +386,13 @@ func TestOptions_CopyRateLimit(t *testing.T) {
 	start := time.Now()
 	err = Copy("test/data/case16", "test/data.copy/case16", opt)
 	elapsed := time.Since(start)
-	Expect(t, err).ToBe(nil)
-	Expect(t, elapsed > 5*time.Second).ToBe(true)
+	if supportsWrapReaderOption {
```
Comment (otiai10) [inline all_test.go:389]: I totally disagree with this `if`.
If I were a user reading this test code,

1. first I would think "hmm, what's this `supportsWrapReaderOption`?"
2. second I would search `supportsWrapReaderOption` and find `setupFileCopyMethod()` and would think "now, what's `TEST_FILECOPYMETHOD`?"
3. third, I would find `${{ matrix.os.copymethod }}` then I would know it's deeply coupled with filesystem

It is a long journey. We need to find more straightforward way so that users can easily understand.

File: .github/workflows/filecopymethod-test.yml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    - name: Copy files to mounted filesystem
+      if: ${{ matrix.environment.filesystem != '' }}
+      run: |-
+        rsync -av --exclude=".*" --exclude "test/filesystems" . "$TEST_PATH"
+
+    - name: Test
+      working-directory: ${{ env.TEST_PATH }}
+      env:
+        TEST_FILESYSTEM: ${{ matrix.environment.filesystem }}
+        TEST_FILECOPYMETHOD: ${{ matrix.os.copymethod }}
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:87]: might be typo of `environment`, not `os`

File: .github/workflows/filecopymethod-test.yml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    runs-on: ${{ matrix.environment.runner }}
+    strategy:
+      matrix:
+        environment:
+         - runner: macos-latest
+           filesystem: APFS
+           copymethod: GetBytes
+         - runner: ubuntu-latest
+           filesystem: btrfs
+           copymethod: GetBytes
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:29]: might be typo of `CopyByte`, not `GetBytes`

File: .github/workflows/filecopymethod-test.yml [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    steps:
+    - name: Set up Go
+      uses: actions/setup-go@v4
+      with:
+        go-version: 'stable'
+      id: go
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:35]: Do we need this`id`

File: .github/workflows/filecopymethod-test.yml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+    - # Sets TEST_PATH environment variable.
+      # TEST_PATH will later be updated to the mountpoint of the filesystem mount point.
+      name: Set up testing environment
+      run: |-
+        mkdir ./test/filesystems
+        echo "TEST_PATH=." >> $GITHUB_ENV
+
+    - name: Set up filesystem (MacOS)
+      if: ${{ matrix.environment.filesystem != '' && startsWith(matrix.environment.runner, 'macos-') }}
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:54]: Please do not use `if` so many times.
It's much far better and easier to separate jobs and `runs_on`

File: .github/workflows/filecopymethod-test.yml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      run: |-
+        IMAGE_PATH="./test/filesystems/${{matrix.environment.filesystem}}.dmg"
+        MOUNT_PATH="./test/filesystems/${{matrix.environment.filesystem}}.mount"
+        echo "TEST_PATH=${MOUNT_PATH}" >> $GITHUB_ENV
+
+        hdiutil create -size 500m -fs APFS "$IMAGE_PATH"
+        hdiutil attach -mountpoint "$MOUNT_PATH" "$IMAGE_PATH"
+
+    - name: Set up filesystem (Linux)
+      if: ${{ matrix.environment.filesystem != '' && startsWith(matrix.environment.runner, 'ubuntu-') }}
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:64]: So repeated but slightly different `if` takes high cognitive load.
It's much far better and easier to separate jobs and `runs_on`

File: .github/workflows/filecopymethod-test.yml [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      run: |-
+        IMAGE_PATH="./test/filesystems/${{matrix.environment.filesystem}}.img"
+        MOUNT_PATH="./test/filesystems/${{matrix.environment.filesystem}}.mount"
+        echo "TEST_PATH=${MOUNT_PATH}" >> $GITHUB_ENV
+
+        truncate -s 500m "$IMAGE_PATH"
+        mkfs -t "${{matrix.environment.filesystem}}" "$IMAGE_PATH"
+        mkdir "$MOUNT_PATH"
+        whoami
+        id -u
```
Comment (otiai10) [inline .github/workflows/filecopymethod-test.yml:74]: debug code?

**Review by otiai10 — CHANGES_REQUESTED** [review]

File: options.go [hunk 14]
```diff
@@ -134,6 +134,10 @@ type FileCopyMethod struct {
 	fcopy func(src, dest string, info os.FileInfo, opt Options) (err error, skipFile bool)
 }
 
+// The default FileCopyMethod.
+// This only is changed during tests.
+var defaultCopyMethod = CopyBytes
```
[the remark this answers is not in the document]
Comment (eth-p) [reply options.go:139]: > Users can set their opt.FileCopyMethod by theirselves, NOT by env or any other implicit mechanism.

The users of `copy` do not have any way to change this as a default. It can only be changed by developers of `copy`, and is only ever changed by the `all_test.go` code.

It's test-specific code, and I agree that's not ideal. My earlier comments explained why it's needed, though.

Maybe naming it `default` is misleading? I could change it to `unspecifiedFileCopyMethodOverride`

File: all_test.go (hunk 7 above)
Comment (eth-p) [reply all_test.go:389]: I could move `supportsWrapReadOption` into the `FileCopyMethod` struct? That might be better.

File: .github/workflows/filecopymethod-test.yml (hunk 8 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml:87]: Good catch, thanks.

File: .github/workflows/filecopymethod-test.yml (hunk 9 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml:29]: Same

File: .github/workflows/filecopymethod-test.yml (hunk 11 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml:54]: Good point. Now that it's in a different workflow, we can do that easily

File: .github/workflows/filecopymethod-test.yml (hunk 13 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml:74]: Good catch, yeah

File: options.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (eth-p) [reply options.go]: I ended up removing this. I replaced it with a function that runs at the end of `getDefaultOptions` and replaces the `FileCopyMethod`.

File: .github/workflows/filecopymethod-test.yml (hunk 10 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml]: I removed it.

File: .github/workflows/filecopymethod-test.yml (hunk 12 above)
Comment (eth-p) [reply .github/workflows/filecopymethod-test.yml]: Done.

**Review by otiai10 — CHANGES_REQUESTED** [review]

**Comment by otiai10** [comment]
Thank you for your input. I understand the points of conflict. I believe that DRY (Don't Repeat Yourself) principles are not as critical in testing. Please keep the code simple. Additionally, this pull request does not seem to address a specific issue. Therefore, I suggest we focus on #166 to solve a real problem. During the process of #166, I may or may not see the benefits of this pull request (#165).

At this moment, I’m not inclined to add logic that makes the test code overly DRY, from a readability perspective. I will not merge this.
